### PR TITLE
Alerting: Alert rules search improvements

### LINKED ISSUE #55175: Alerting: Search improvements
The current search feature is a bit cumbersome to use; it's very particular about the syntax and we could benefit from fuzzy searching or free form text search.

We can also (perhaps as a stretch goal) make it easier to create a correct search syntax and add access to simple quick search values.

### LINKED ISSUE #56875: Labels' values containing commas are not working
We support two ways of searching by labels:
- `key=value`
- `{key="value"}`
Maybe we should stick to only one (probably the latter as it allows supporting commas). 
However, typing without brackets and quotes is faster

### PR DESCRIPTION
**Pull request by konrad147**
**What is this feature?**
This PR adds basic search query syntax to the alert rules list.

**Why do we need this feature?**
The current search allows only filtering rules by their names and assigned labels.

However, alert rules are organized in a hierarchical structure. 
- Folder / Namespace
	- Evaluation group
	  - Alert rule 1
	  - Alert rule 2

To effectively search in a structure like this, we often need to define multiple filtering criteria on different levels of nesting.
To support that scenario, we created a simple search syntax that allows applying different sets of filters in a single query.
Thanks to the search syntax, we can apply filtering on different levels of the structure, like folders, evaluation groups, and rules. Additionally, we can filter rules by their properties like labels, state, type, and health

The list of supported filters is as follows:
| Filtered entity | Expression|
|---|---|
| Data Source | `datasource:datasource_name`|
| Folder or Namespace | `namespace:namespace_name`|
| Group | `group:group_name` | 
| Rule | `rule:rule_name` |
| Labels | `label:foo=bar` `label:foo=baz` |
| State | `state:firing/normal/pending` |
| Type | `type:alerting/recording` |
| Health | `health:ok/nodata/error` |

**Which issue(s) does this PR fix?**:
Fixes #55175 #56875

**Special notes for your reviewer**:
This PR introduces Github-inspired search syntax for the alert rules list
The search supports filtering by data source, namespace/folder, rule name, group name, labels, alert state, and alert type.

It also includes a compatibility layer that translates the old query syntax into the new one, so the old URLs containing search queries should work.

**DEMO**

https://user-images.githubusercontent.com/868844/215752093-ada54022-834e-40f0-980b-dcf882b75626.mp4

**Screenshot**
image

### CODE REVIEW

**Comment by gillesdemey** [comment]
Can we add some instructions to the project as a README on how to compile the grammar?

**Review by soniaAguilarPeiron — APPROVED** [review]
Amazing work!!! 🎉, not only because this will be a super useful feature for the users, but for the way you implemented it  🙌
I added a couple of nits, no blockers 🚀

File: public/app/features/alerting/unified/search/search.grammar [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ DataSourceFilter |
+ NameSpaceFilter |
+ LabelFilter |
+ GroupFilter |
+ RuleFilter |
+ StateFilter |
+ TypeFilter
+}
+
+DataSourceFilter { filter<"ds" | "datasource"> }
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/search/search.grammar:15]: I don't think we should support these short-hand aliases and might add an additional maintenance burden. 

A few reasons I've mention in our review call;

* It assumes readers of the query know the keyword the short-hand refers to
* It adds additional code maintenance for them in the search code
* Adds additional complexity when we want to add additional features
	(ie. if we wanted to implement some sort of "type-ahead", we'd also have to make it work for these and the single character letters could add additional complexity to the code)

File: public/app/features/alerting/unified/components/rules/RulesFilter.tsx [hunk 2]
```diff
@@ -58,155 +62,183 @@ const RulesFilter = () => {
...
   const handleClearFiltersClick = () => {
-    setQueryParams({
-      alertState: null,
-      queryString: null,
-      dataSource: null,
-      ruleType: null,
-    });
+    setSearchQuery(undefined);
+
+    if (onFilterCleared) {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/RulesFilter.tsx:97]: Perhaps we should default this to a `noop` instead of checking if it's defined?

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts [hunk 3]
```diff
@@ -1,35 +1,103 @@
...
-  return useMemo(() => {
-    const filteredNamespaces = namespaces
-      // Filter by data source
-      // TODO: filter by multiple data sources for grafana-managed alerts
+export function useRulesFilter() {
+  const [queryParams, updateQueryParams] = useURLSearchParams();
+  const searchQuery = queryParams.get('search') ?? '';
+
+  const filterState = getSearchFilterFromQuery(searchQuery);
+  const hasActiveFilters = Object.values(filterState).some((fs) => (Array.isArray(fs) ? fs.length > 0 : !!fs));
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/hooks/useFilteredRules.ts:22]: I think we can shorten this to `Boolean(fs?.length)`? Can it be something other than `string` or `Array`?

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts [hunk 4]
```diff
@@ -1,35 +1,103 @@
...
+  // Handle legacy filters
+  useEffect(() => {
+    const legacyFilters = {
+      dataSource: queryParams.get('dataSource') ?? undefined,
+      alertState: queryParams.get('alertState') ?? undefined,
+      ruleType: queryParams.get('ruleType') ?? undefined,
+      labels: parseMatchers(queryParams.get('queryString') ?? '').map(matcherToMatcherField),
+    };
+
+    const hasLegacyFilters = Object.values(legacyFilters).some((lf) => (Array.isArray(lf) ? lf.length > 0 : !!lf));
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/hooks/useFilteredRules.ts:48]: Dito shorter check without ternary as above

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts [hunk 5]
```diff
@@ -103,27 +89,35 @@ export const useFilteredRules = (namespaces: CombinedRuleNamespace[]) => {
...
+      .filter((ns) => (ngFilters.namespace ? ns.name.toLowerCase().includes(ngFilters.namespace.toLowerCase()) : true))
       .filter(({ rulesSource }) =>
-        filters.dataSource && isCloudRulesSource(rulesSource) ? rulesSource.name === filters.dataSource : true
+        ngFilters.dataSourceName && isCloudRulesSource(rulesSource)
+          ? rulesSource.name === ngFilters.dataSourceName
+          : true
       )
       // If a namespace and group have rules that match the rules filters then keep them.
-      .reduce(reduceNamespaces(filters), [] as CombinedRuleNamespace[])
+      .reduce(reduceNamespaces(ngFilters), [] as CombinedRuleNamespace[])
```
Comment (VikaCep) [inline public/app/features/alerting/unified/hooks/useFilteredRules.ts:92]: why is it called `ngFilters`? `ng` takes me back to angular 😞

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts [hunk 6]
```diff
@@ -43,35 +111,52 @@ const reduceNamespaces = (filters: FilterState) => {
...
   return (groupAcc: CombinedRuleGroup[], group: CombinedRuleGroup) => {
     const rules = group.rules.filter((rule) => {
-      if (filters.ruleType && filters.ruleType !== rule.promRule?.type) {
+      if (ngFilters.ruleType && ngFilters.ruleType !== rule.promRule?.type) {
+        return false;
+      }
+      if (
+        ngFilters.dataSourceName &&
+        isGrafanaRulerRule(rule.rulerRule) &&
+        !isQueryingDataSource(rule.rulerRule, ngFilters)
```
Comment (VikaCep) [inline public/app/features/alerting/unified/hooks/useFilteredRules.ts:123]: What do you think of abstracting this condition into a `const`? it's hard to tell what it checks for at a glance.

File: public/app/features/alerting/unified/components/rules/RulesFilter.tsx [hunk 7]
```diff
@@ -42,14 +42,18 @@ const RuleTypeOptions: SelectableValue[] = [
...
 ];
 
-const RulesFilter = () => {
-  const [queryParams, setQueryParams] = useQueryParams();
+interface RulesFilerProps {
+  onFilterCleared?: () => void;
+}
+
+const RulesFilter = ({ onFilterCleared }: RulesFilerProps) => {
+  const [queryParams, setQueryParams] = useQueryParams(); // TODO Limit scope of query params
```
Comment (VikaCep) [inline public/app/features/alerting/unified/components/rules/RulesFilter.tsx:50]: Is this @todo left there on purpose?

File: public/app/features/alerting/unified/search/searchParser.test.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    it('should parse non-filtering words as free form query', () => {
+      const filter = getSearchFilterFromQuery('cpu usage rule');
+      expect(filter.freeFormWords).toHaveLength(3);
+      expect(filter.freeFormWords).toContain('cpu');
+      expect(filter.freeFormWords).toContain('usage');
+      expect(filter.freeFormWords).toContain('rule');
+    });
+
+    it('should parse filter values with whitespaces when in quotes', () => {
```
Comment (VikaCep) [inline public/app/features/alerting/unified/search/searchParser.test.ts:56]: Would it also work with single quotes?

File: public/app/features/alerting/unified/components/rules/RulesFilter.tsx [hunk 9]
```diff
@@ -159,14 +159,7 @@ const RulesFilter = ({ onFilterCleared }: RulesFilerProps) => {
...
           </Stack>
-          {(filterState.freeFormWords.length ||
-            filterState.dataSourceName ||
-            filterState.namespace ||
-            filterState.ruleType ||
-            filterState.ruleState ||
-            filterState.labels.length ||
-            filterState.groupName ||
-            filterState.ruleName) && (
+          {hasActiveFilters && (
```
Comment (VikaCep) [inline public/app/features/alerting/unified/components/rules/RulesFilter.tsx:162]: Nice!

**Review by VikaCep — COMMENTED** [review]
Awesome work! 👏 👏 👏

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts (hunk 3 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:22]: what about using the `isEmpty` from Lodash?

File: public/app/features/alerting/unified/search/searchParser.test.ts (hunk 8 above)
Comment (konrad147) [reply public/app/features/alerting/unified/search/searchParser.test.ts:56]: No

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts (hunk 3 above)
Comment (konrad147) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:22]: Good ideas!

File: public/app/features/alerting/unified/RuleList.tsx [hunk 10]
```diff
@@ -83,19 +81,20 @@ const RuleList = withErrorBoundary(
...
-    const filteredNamespaces = useFilteredRules(combinedNamespaces);
+    const filteredNamespaces = useFilteredRules(combinedNamespaces, filterState);
+
     return (
       // We don't want to show the Loading... indicator for the whole page.
       // We show separate indicators for Grafana-managed and Cloud rules
       <AlertingPageWrapper pageId="alert-list" isLoading={false}>
         <RuleListErrors />
-        <RulesFilter />
+        <RulesFilter onFilterCleared={() => setExpandAll(false)} />
```
[the remark this answers is not in the document]
Comment (konrad147) [reply public/app/features/alerting/unified/RuleList.tsx:91]: It's a bugfix. Currently, we only show the `Expand`/`Collapse` buttons if any filters are applied. If you expand and then clear filters, then you end up with an expanded list of all alerts, and there is no way to collapse it back

File: public/app/features/alerting/unified/components/rules/RulesFilter.tsx (hunk 7 above)
Comment (konrad147) [reply public/app/features/alerting/unified/components/rules/RulesFilter.tsx:50]: No, I'll remove

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts (hunk 6 above)
Comment (konrad147) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:123]: I'll add this. In general, I'd like to refactor the filtering logic, but it's a pretty complex task, and I don't want to include it in this PR

File: public/app/features/alerting/unified/components/rules/RulesFilter.tsx (hunk 2 above)
Comment (konrad147) [reply public/app/features/alerting/unified/components/rules/RulesFilter.tsx:97]: It's a prop for an additional action that you may want to invoke after clearing out filters not to clear the filter expression

**Comment by gillesdemey** [comment]
Added a test for quoted text search - LGTM no blockers!

**Review by gillesdemey — APPROVED** [review]

**Review by VikaCep — APPROVED** [review]
👏 👏 👏

**Comment by brendamuir** [comment]
Two mini ui text suggestions -- can we rename the button to +Create alert rule - makes it more actionable and it's already clear that it's new. 
Also for the description text - how about making it actionable too - View a list of alert rules and their status"

**Review by brendamuir — APPROVED** [review]
